tests/containers/containers_test.go

```diff
@@ -22,11 +22,56 @@ import (
 	"github.com/stretchr/testify/assert"
 
 	ptesting "github.com/pulumi/pulumi/pkg/testing"
+	"github.com/pulumi/pulumi/pkg/testing/integration"
 )
 
-// TestPulumiContainerImages simulates building and running Pulumi programs
-// on all of the supported container images.
-func TestPulumiContainerImages(t *testing.T) {
+// TestPulumiDockerImage simulates building and running Pulumi programs on the pulumi/pulumi Docker image.
+//
+// NOTE: this test is intended to be run inside the aforementioned container, unlike the actions test below.
+func TestPulumiDockerImage(t *testing.T) {
+	const stackOwner = "moolumi"
+
+	if os.Getenv("RUN_CONTAINER_TESTS") == "" {
+		t.Skip("Skipping container runtime tests because RUN_CONTAINER_TESTS not set.")
+	}
+
+	// Confirm we have credentials.
+	if os.Getenv("PULUMI_ACCESS_TOKEN") == "" {
+		t.Fatal("PULUMI_ACCESS_TOKEN not found, aborting tests.")
+	}
+
+	base := integration.ProgramTestOptions{
+		Tracing:              "https://tracing.pulumi-engineering.com/collector/api/v1/spans",
+		ExpectRefreshChanges: true,
+		Quick:                true,
+		SkipRefresh:          true,
+		NoParallel:           true, // we mark tests as Parallel manually when instantiating
+	}
+
+	for _, template := range []string{"csharp", "python", "typescript"} {
+		t.Run(template, func(t *testing.T) {
+			t.Parallel()
+
+			e := ptesting.NewEnvironment(t)
+			defer func() {
+				e.RunCommand("pulumi", "stack", "rm", "--force", "--yes")
+				e.DeleteEnvironment()
+			}()
+
+			stackName := fmt.Sprintf("%s/container-%s-%x", stackOwner, template, time.Now().UnixNano())
+			e.RunCommand("pulumi", "new", template, "-f", "-s", stackName)
+
+			example := base.With(integration.ProgramTestOptions{
+				Dir: e.RootPath,
+			})
+
+			integration.ProgramTest(t, &example)
+		})
+	}
+}
+
+// TestPulumiActionsImage simulates building and running Pulumi programs on the pulumi/actions image.
+func TestPulumiActionsImage(t *testing.T) {
```

Comment:
It appears we are never running this test - is that right/intentional?

Reply:
I'll admit this feels a little weird now too - we have some tests that must run outside and some that must run inside.  I don't exactly follow why we wanted these to diverge.

Reply:
@lukehoban yes, it was intentional.

The `pulumi/pulumi` tests run _within_ the Docker container, and go through the prototypical `pulumi stack init`, `pulumi up`, ... set of scenarios. Those tests are to validate that Pulumi does all the things it needs to do within that container runtime.

The `pulumi/actions` tests cover a different use case. `pulumi/actions` is used for GitHub Actions, and is used for things like CI. So the most common use-case is simply passing the args "up --stack foo/bar". And then the entry point in the Dockerfile will run `yarn install` or `pip`, etc. to get the right dependencies and then run `pulumi`.

So specifically we want those tests to exercise the entry point for the docker container itself.

It's complicated and should be called out in some long comments, but that's the idea anyway.

For expediency Pat went and added just this set of tests. I'll send out a PR to hook up the `pulumi/actions` container, running outside of the Docker container soonish.